# Lecture 14 - Reading and Writing files

We have already talked about built-in Python types, but there are more types that we did not speak about. One of these is the ``file()`` object which can be used to read or write files.

## Reading files

Let's try and get the contents of the file into IPython. We start off by creating a file object:

In [17]:
f = open('data1.txt', 'r')

The ``open`` function is taking the [data1.txt](data1.txt) file, opening it, and returning an object (which we call ``f``) that can then be used to access the data.

Note that ``f`` is not the data in the file, it is what is called a *file handle*, which points to the file:

In [19]:
type(f)

_io.TextIOWrapper

Now, simply type:

In [21]:
f.read()

'Latitude   Logitude   Identifier        Alt    e_Alt\n(deg)      (deg)                        (km)   (km) \n---------- ---------- ----------------- ------ ------\n010.684737 +41.269035 00424433+4116085   1.453  0.052\n010.683469 +41.268585 00424403+4116069   1.321  0.022\n010.685657 +41.269550 00424455+4116103   2.773  0.069\n010.686026 +41.269226 00424464+4116092   1.299  0.063\n010.683465 +41.269676 00424403+4116108   2.507  0.056\n010.686015 +41.269630 00424464+4116106   1.399  0.045\n010.685270 +41.267124 00424446+4116016   2.070  0.035\n'

The ``read()`` function basically just read the whole file and put the contents inside a string.

Let's try this again:

In [23]:
f.read()

''

What's happened? We read the file, and the file 'pointer' is now sitting at the end of the file, and there is nothing left to read.

Let's now try and do something more useful, and capture the contents of the file in a string:

In [3]:
f = open('data1.txt', 'r')
data = f.read()

Now ``data`` should contain a string with the contents of the file:

In [5]:
data

'Latitude   Logitude   Identifier        Alt    e_Alt\n(deg)      (deg)                        (km)   (km) \n---------- ---------- ----------------- ------ ------\n010.684737 +41.269035 00424433+4116085   1.453  0.052\n010.683469 +41.268585 00424403+4116069   1.321  0.022\n010.685657 +41.269550 00424455+4116103   2.773  0.069\n010.686026 +41.269226 00424464+4116092   1.299  0.063\n010.683465 +41.269676 00424403+4116108   2.507  0.056\n010.686015 +41.269630 00424464+4116106   1.399  0.045\n010.685270 +41.267124 00424446+4116016   2.070  0.035\n'

But what we'd really like to do is read the file line by line. There are several ways to do this, the simplest of which is to use a ``for`` loop in the following way:

In [9]:
f = open('data1.txt', 'r')
for line in f:
    print(repr(line))

'Latitude   Logitude   Identifier        Alt    e_Alt\n'
'(deg)      (deg)                        (km)   (km) \n'
'---------- ---------- ----------------- ------ ------\n'
'010.684737 +41.269035 00424433+4116085   1.453  0.052\n'
'010.683469 +41.268585 00424403+4116069   1.321  0.022\n'
'010.685657 +41.269550 00424455+4116103   2.773  0.069\n'
'010.686026 +41.269226 00424464+4116092   1.299  0.063\n'
'010.683465 +41.269676 00424403+4116108   2.507  0.056\n'
'010.686015 +41.269630 00424464+4116106   1.399  0.045\n'
'010.685270 +41.267124 00424446+4116016   2.070  0.035\n'


Note that we are using ``repr()`` to show any invisible characters (this will be useful in a minute). Also note that we are now looping over a file rather than a list, and this automatically reads in the next line at each iteration. Each line is being returned as a string. Notice the ``\n`` at the end of each line - this is a line return character, which indicates the end of a line.

Now we're reading in a file line by line, what would be nice would be to get some values out of it.  Let's examine the last line in detail. If we just type ``line`` we should see the last line that was printed in the loop:

In [11]:
line

'010.685270 +41.267124 00424446+4116016   2.070  0.035\n'

We can first get rid of the ``\n`` character with:

In [13]:
line = line.strip()

In [15]:
line

'010.685270 +41.267124 00424446+4116016   2.070  0.035'

Next, we can use what we learned about strings and lists to do:

In [17]:
columns = line.split()

In [19]:
columns

['010.685270', '+41.267124', '00424446+4116016', '2.070', '0.035']

Finally, let's say we care about the mountain name (the Location column), and the Altitude (the Alt) column:

In [21]:
name = columns[2]
alt = columns[3]

In [23]:
name

'00424446+4116016'

In [47]:
alt

'2.070'

Note that ``alt`` is a string, but if we want a floating point number, we can instead do:

In [49]:
alt = float(columns[3])

In [51]:
alt

2.07

One last piece of information we need about files is how we can read a single line. This is done using:

    line = f.readline()

We can put all this together to write a little script to read the data from the file and display the columns we care about to the screen! Here is is:

In [57]:
# Open file
f = open('data1.txt', 'r')

# Read and ignore header lines
header1 = f.readline()
header2 = f.readline()
header3 = f.readline()

# Loop over lines and extract variables of interest
for line in f:
    line = line.strip()
    columns = line.split()
    name = columns[2]
    alt = float(columns[3])
    print(name, alt)

00424433+4116085 1.453
00424403+4116069 1.321
00424455+4116103 2.773
00424464+4116092 1.299
00424403+4116108 2.507
00424464+4116106 1.399
00424446+4116016 2.07


## Exercise 1

Here is a copy of the above code to read in the file. Modify this code so as to create a dictionary which gives ``Alt`` for a given ``Location`` name, i.e.

    >>> alt['00424455+4116103']
    10.773
    
Then loop over the items in the dictionary and print out for each the source name and the ``alt`` value.

In [61]:
# EDIT THE CODE BELOW

# Open file
f = open('data1.txt', 'r')

# Read and ignore header lines
header1 = f.readline()
header2 = f.readline()
header3 = f.readline()

# Loop over lines and extract variables of interest
for line in f:
    line = line.strip()
    columns = line.split()
    name = columns[2]
    alt = float(columns[3])
    print(name, alt)
    
    

00424433+4116085 1.453
00424403+4116069 1.321
00424455+4116103 2.773
00424464+4116092 1.299
00424403+4116108 2.507
00424464+4116106 1.399
00424446+4116016 2.07


**Bonus:** can you figure out a way to make sure that you loop over the source names in alphabetical order?

## Writing files

To open a file for writing, use:

In [63]:
f = open('data_new.txt', 'w')

Then simply use ``f.write()`` to write any content to the file, for example:

In [65]:
f.write("Hello, World!\n")

14

If you want to write multiple lines, you can either give a list of strings to the ``writelines()`` method:

In [67]:
f.writelines(['spam\n', 'egg\n', 'spam\n'])

or you can write them as a single string:

In [69]:
f.write('spam\negg\nspam\n')

14

Once you have finished writing data to a file, you need to close it:

In [71]:
f.close()

(this also applies to reading files)

## The with-statement

As we have seen above, files must not just be opened but should be properly closed afterwards to make sure they are actually written before using them somewhere else. Sometimes writes to files get cached by Python to minimize actual writing to disk, which is comparably slow. Closing a file ensures that these changes are actually written.

To avoid forgetting to close a file there is the with-statement.

In [73]:
with open('data_new.txt', 'w') as f:
    f.write('spam\n')

This opens the specified file and holds the file-object within ``f``, as well as closing the file when the with-codeblock ends. Afterwards, the file is properly closed and not available anymore.

## Exercise 2

Continuing from the example in the 'Reading files' section, can you figure out a way to write out a new file containing two columns - the name of the source and the ``alt`` value?

In [ ]:
# EDIT THE CODE BELOW

# Open file
f = open('data1.txt', 'r')

# Read and ignore header lines
header1 = f.readline()
header2 = f.readline()
header3 = f.readline()

# Loop over lines and extract variables of interest
for line in f:
    line = line.strip()
    columns = line.split()
    name = columns[2]
    alt = float(columns[3])

## Notes

The above shows you how you can read and write any data file. Of course, there are also functions that exist to help you read in data in certain formats (for example ``numpy`` contains a function ``numpy.loadtxt`` to read in arrays from files) but the key is that with the above, you can read any file, so any other function is then just making your life easier.

## Homework #6 

**REMINDER:** *All coding assignment will be turned in as .ipynb files, to the same PHYS_X0223 repository on GitHub.*   
*They should be turned in with the following naming:*
    
    Lastname_Firstinitial_24_HW6.ipynb
    

We learned basic error propigations in HW1.  

Read in the data from 'data1.txt'.

Write a **function** that will compute, what you did by hand in HW1, the mean and standard deviation of the 'alt' from the data points in the file we read in (data1.txt).

Reminder:

$$a = \frac{\sum_{i=1}^N (a_i/\sigma_i^2)}{\sum_{i=1}^N (1 / \sigma_i^2) } \;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\; 1/\sigma^2 =  \sum_{i=1}^N (1 / \sigma_i^2)$$

where $a_i = alt$  and $\sigma_i = e\_alt$

**Write a code that will (1) read the file, (2) use the fuction, (3) print the results.**

In [73]:
import numpy as np
def mean_error(a_i, sigma_i):
    x = [1/i**2 for i in sigma_i] #create new list with values of 1 over sigma squared, equation provided
    y = sum(a_i[i]*x[i] for i in range(6)) #range 6 (including 0) due to 7 data points for each column
    a = y/sum(x) #finish the calculation
    sigma = (1/sum(x))**0.5 #calculate sigma, equation provided above
    return a,sigma
    
a_i=[] #empty list to be populated
sigma_i=[] #empty list to be populated

# read in file
f = open('data1.txt', 'r')
#data = f.read()  not sure what to do with this provided line so I removed it
header1 = f.readline() #removing top three rows to isolate data
header2 = f.readline()
header3 = f.readline()

for line in f:
    line = line.strip()
    columns = line.split()
    alt = float(columns[3]) #isolating alt data
    e_alt = float(columns[4]) #isolating e_alt data
    a_i.append(alt) #adding alt data row by row
    sigma_i.append(e_alt) #adding e_alt data row by row
    
# call function
a, sigma = mean_error(a_i, sigma_i)
# print results
print("the mean equals",a)
print("the standard deviation equals",sigma)


the mean equals 1.2517684714311557
the standard deviation equals 0.014862661163658227
